### PR TITLE
feat(manager/gomod): add `goGetDirs` option

### PR DESCRIPTION
**Pull request by jorgemmsilva**
## Changes

adds the `goGetDirs` option to allow customization of `go get`

## Context

fixes the issue described in:
https://github.com/renovatebot/renovate/discussions/20066

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/gomod/artifacts.ts
Comment (rarkins) [inline lib/modules/manager/gomod/artifacts.ts:275]: Did you find that each path needs to be quoted individually instead of collectively?

We need to check whether it's possible to bypass this quoting by configuring like `["./...", "&&", "cat", "/etc/passwd"]`

**Review by rarkins — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (viceice) [inline .gitignore:21]:
```suggestion
```
no

File: lib/modules/manager/gomod/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.spec.ts:1590]: use partial function from test utils

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/gomod/artifacts.ts
Comment (jorgemmsilva) [reply lib/modules/manager/gomod/artifacts.ts:275]:
```
goGetDirs: ['.', 'foo', '.bar/...', '&&', 'cat', '/etc/passwd']
```
will result in 
```
go get -d -t . foo .bar/... '&&' cat /etc/passwd
```
which will crash and not allow the command chaining.

this is what is expected, right?

File: lib/modules/manager/gomod/artifacts.spec.ts
Comment (jorgemmsilva) [reply lib/modules/manager/gomod/artifacts.spec.ts]: sure, I updated to use `partial`, but all the other tests in the same file don't use it :shrug:

File: lib/modules/manager/gomod/artifacts.ts
Comment (rarkins) [reply lib/modules/manager/gomod/artifacts.ts:275]: Paging @secustor and @JamieMagee for additional opinions

Comment (secustor) [reply lib/modules/manager/gomod/artifacts.ts:275]: Not sure if this can be defeated by using a combination of single and double quotes as well backslashes. :/

We could check the strings with [`path.parse(path)`](https://nodejs.org/docs/latest-v18.x/api/path.html#pathparsepath) and add one exception for `./...`

Comment (rarkins) [reply lib/modules/manager/gomod/artifacts.ts:275]: This user wants it configurable though

Comment (viceice) [reply lib/modules/manager/gomod/artifacts.ts:275]: this allows to pass arbitrary ARGS to go get. so we somehow need to defend it

Comment (jorgemmsilva) [reply lib/modules/manager/gomod/artifacts.ts:275]: >Not sure if this can be defeated by using a combination of single and double quotes as well backslashes. :/

I mean, it would be easy to remove quotes and backslashes from the input strings, those are not valid filepaths anyways :shrug:

Comment (secustor) [reply lib/modules/manager/gomod/artifacts.ts:275]: Yes, that should be good enough for now. 

Create a utility function for this in https://github.com/renovatebot/renovate/blob/main/lib/util/fs/index.ts and name it `isValidPath`.

Comment (jorgemmsilva) [reply lib/modules/manager/gomod/artifacts.ts:275]: well, I tried adding such functionality, but that module is mocked in the tests:
```
jest.mock('../../../util/fs',);
```

I tried using `jest.requireActual` to restore functionality but to no avail.
What would be the proper way to "unmock" that function? 

Give me a discount as I have not touched JS in like 2 years :smile:

Comment (secustor) [reply lib/modules/manager/gomod/artifacts.ts]: This blog describes how you can setup a mock like you are trying to do: https://www.benmvp.com/blog/mock-all-functions-module-except-one-jest/

But I'm fine with mocking the result of `isValidPath` for gomod, if the examples from your tests are included in the ones for the utility function.

File: lib/modules/manager/gomod/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.spec.ts:1614]: forbid absolute paths

File: lib/util/fs/index.spec.ts
Comment (viceice) [inline lib/util/fs/index.spec.ts:505]: use `it.each` jest table syntax here

File: lib/util/fs/index.ts
Comment (viceice) [inline lib/util/fs/index.ts:305]:
```suggestion
export function isValidLocalPath(s: string): boolean {
```
disallow absolute paths, so no leading slash.

we should also disallow `./../` to not going outside of the repo.

i think we already have a function to validate that a path is inside the repo .

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/gomod/artifacts.spec.ts
Comment (secustor) [inline lib/modules/manager/gomod/artifacts.spec.ts:1614]: Please add a test case where actual files have been upgraded.    
It is enough to copy an existing happy path case and add your new option.

File: lib/modules/manager/gomod/artifacts.ts
Comment (secustor) [inline lib/modules/manager/gomod/artifacts.ts:280]:
```suggestion
    let args = `get -d -t ${goGetDirs}`;
```

File: lib/modules/manager/gomod/artifacts.ts
Comment (secustor) [inline lib/modules/manager/gomod/artifacts.ts:276]: Please add a debug log which informs the users what parts have been stripped, else they will be confused what is happening.

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/gomod/artifacts.spec.ts
Comment (jorgemmsilva) [reply lib/modules/manager/gomod/artifacts.spec.ts:1614]: could you please point to an example test that already does that?

Comment (secustor) [reply lib/modules/manager/gomod/artifacts.spec.ts:1614]: https://github.com/renovatebot/renovate/blob/e6ebf1dc16a78872a2b30cbf11c33dd20d84e71f/lib/modules/manager/gomod/artifacts.spec.ts#L134-L178

**Review by secustor — DISMISSED** [review]

File: lib/modules/manager/gomod/artifacts.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.ts:280]: @rarkins i think we should log that as metadata?

File: lib/modules/manager/gomod/artifacts.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.ts:276]: please make this fully conditional inside a single `if`, so we don't need so many conditional statements

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/gomod/artifacts.ts
Comment (jorgemmsilva) [reply lib/modules/manager/gomod/artifacts.ts:280]: what do you mean by "log as metadata"?

Comment (rarkins) [reply lib/modules/manager/gomod/artifacts.ts]: like `logger.debug({ userGoGetDirs }, 'Blah blah')`

File: docs/usage/configuration-options.md
Comment (rarkins) [inline docs/usage/configuration-options.md:915]: Let's have an example below here

File: lib/modules/manager/gomod/artifacts.ts
Comment (rarkins) [inline lib/modules/manager/gomod/artifacts.ts:283]:
```suggestion
              logger.warn({ dir }, 'Invalid path in goGetDirs';
```

File: lib/util/fs/index.ts
Comment (rarkins) [inline lib/util/fs/index.ts:305]: When I look at this regex, it's very hard to decide "is this right? is this safe?"

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/manager/gomod/artifacts.ts
Comment (viceice) [reply lib/modules/manager/gomod/artifacts.ts:283]: dir should be inside message, otherwise it's not visible on dashboard. 🤔

File: lib/util/fs/index.ts
Comment (jorgemmsilva) [reply lib/util/fs/index.ts:305]: sure, I'll break it down:
```
^                       // sting start
(?!\/)                 // does not have a leading slash '/'
(\.{1})?             // may or may not have a leading dot '.'
([a-z0-9/]+)?   // an arbitrary amount of alphanumeric chars, or slash `/`
(\/\.{3})?         // may or may not finish with triple dot '...'
$                    // string end
```
Hope that's clearer. I can add a description like this to the code if that helps.

Should be  safe, at least I couldn't think of a way to break it.

File: lib/modules/manager/gomod/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.spec.ts:25]:
```suggestion
    ...(jest.createMockFromModule('../../../util/fs')),
    isValidLocalPath: jest.requireActual('../../../util/fs')
```
doesn't `requireActual` return any?

File: lib/modules/manager/gomod/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.spec.ts:1849]: use partial like above

File: lib/modules/manager/gomod/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.spec.ts:1829]: remove pptions, as it's not relevant for this test

File: lib/modules/manager/gomod/artifacts.spec.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.spec.ts:1874]: same here

File: lib/modules/manager/gomod/artifacts.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.ts:276]:
```suggestion
    let goGetDirs: string | undefined;
```

File: lib/modules/manager/gomod/artifacts.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.ts:288]:
```suggestion
          .join(' ');
```

File: lib/modules/manager/gomod/artifacts.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.ts:291]:
```suggestion
    let args = `get -d -t ${goGetDirs ?? './...'}`;
```

File: lib/util/fs/index.ts
Comment (viceice) [inline lib/util/fs/index.ts:308]: we should call `ensureValidLocalPath` to make function more save.

https://github.com/renovatebot/renovate/blob/cc33d34ca827f2acc5a00d69f38328b26ab06939/lib/util/fs/util.ts#L25

File: lib/modules/manager/gomod/artifacts.spec.ts
Comment (jorgemmsilva) [reply lib/modules/manager/gomod/artifacts.spec.ts:25]: yeah, that works on the assignment to `isValidLocalPath`, but  can't do the spread thing without the "as any" cast:

```
Spread types may only be created from object types.ts(2698)
```

Comment (jorgemmsilva) [reply lib/modules/manager/gomod/artifacts.spec.ts]: this makes the test fail actually:

```
jest.requireActual('../../../util/fs').isValidLocalPath,
```
with:
```
Object is of type 'unknown'.ts(2571)
```

---
this works:
```
 (jest.requireActual('../../../util/fs') as any).isValidLocalPath
```

File: lib/modules/manager/gomod/artifacts.ts
Comment (rarkins) [inline lib/modules/manager/gomod/artifacts.ts:305]:
```suggestion
            logger.warn({ dir }, 'Invalid path in goGetDirs');
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/fs/index.ts
Comment (rarkins) [reply lib/util/fs/index.ts:305]: Maybe add that description as a comment in the code.

Also, any chance you could explain to me how `(\.{1})?` is different from `\.?`?

Comment (jorgemmsilva) [reply lib/util/fs/index.ts]: it's the same thing :upside_down_face: 
updated the PR with the description. Any chance we can get this merged soon?

**Review by rarkins — DISMISSED** [review]

**Review by secustor — DISMISSED** [review]

File: lib/modules/manager/gomod/artifacts.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.ts:311]: please extract a helper function

File: lib/util/fs/index.ts
Comment (viceice) [reply lib/util/fs/index.ts:313]: why this is resolved without comment or code changes?

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/fs/index.ts
Comment (jorgemmsilva) [reply lib/util/fs/index.ts:313]: https://github.com/jorgemmsilva/renovate/blob/feat/gogetdir/lib/modules/manager/gomod/artifacts.ts#L312

File: lib/modules/manager/gomod/artifacts.ts
Comment (jorgemmsilva) [reply lib/modules/manager/gomod/artifacts.ts:311]: That is really a matter of taste, why extract this to a function that is not used anywhere else that pollutes the package?

what would you even call that function? `isValidLocalPathAndLog` ?

To be honest that's the kind of change that as a maintainer you can do after this is merged.

File: lib/util/fs/index.ts
Comment (viceice) [reply lib/util/fs/index.ts:313]: should be inside this function

**Review by viceice — COMMENTED** [review]
see comment

File: lib/util/fs/index.ts
Comment (jorgemmsilva) [reply lib/util/fs/index.ts:313]: different purpose, one is for filtering, another is for transformation.

`isValidLocalPath` - returns whether a given path is valid
 `ensureLocalPath` - returns an absolute path from a local path, I would actually argue that the name of this function is misleading

 I don't understand the benefit of mixing those.

Comment (viceice) [reply lib/util/fs/index.ts:313]: the idea was not to call ensure path, but do the same validation here. that probably needs some refactoring of the util functions.

that is much safer that any regex testing.

Comment (jorgemmsilva) [reply lib/util/fs/index.ts:313]: why not both? I'm having trouble understanding the change you're requesting.
Honestly this has been opened for 2 months. I'm just trying to contribute with a feature that would be useful to me and others. Please just make the change or change things as you see fit

Comment (viceice) [reply lib/util/fs/index.ts:313]: ok, I'll try to add the function as a separate PR, so you only need to rebase after.

Comment (jorgemmsilva) [reply lib/util/fs/index.ts:313]: any update on this @viceice ?

Comment (viceice) [reply lib/util/fs/index.ts:313]: i hopefully can do it tomorrow morning

Comment (viceice) [reply lib/util/fs/index.ts:313]: PR is merged, so please update this PR a hopefully last time.

Comment (jorgemmsilva) [reply lib/util/fs/index.ts:322]: just updated (rebased) the PR, the tests still pass correctly.
Is there any extra change you would expect?

File: lib/modules/manager/gomod/artifacts.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.ts:312]:
```suggestion
```

no longer needed

File: lib/modules/manager/gomod/artifacts.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.ts:18]:
```suggestion
```

see below

File: lib/modules/manager/gomod/artifacts.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.ts:315]: should we check if `goGetDirs===' '` if the filter had returned an empty list?

Comment (jorgemmsilva) [reply lib/modules/manager/gomod/artifacts.ts:315]:
```
[].join(" ")
```
yields `""`

so that shouldn't be necessary

Comment (viceice) [reply lib/modules/manager/gomod/artifacts.ts:315]: but `(goGetDirs ?? './...') === ''`

Comment (jorgemmsilva) [reply lib/modules/manager/gomod/artifacts.ts:315]: sure, I think that's okay.
if an user tries to specify some invalid godirs, it's better to let the command fail, than to "auto-magically fix it" to the default "./..."

Comment (viceice) [reply lib/modules/manager/gomod/artifacts.ts:315]: i would prefer to return an artifact error early to save time. so we don't install and run go if we know it fails anyways. we should not waste time

Comment (jorgemmsilva) [reply lib/modules/manager/gomod/artifacts.ts:315]: would something like this suffice?
```
    if (goGetDirs === '') {
      throw new Error('Invalid goGetDirs');
    }
```

**Review by viceice — APPROVED** [review]

**Review by secustor — APPROVED** [review]

**Pull request merged**
